<a href="https://colab.research.google.com/github/Cedrik-Black/Data_science/blob/master/exercises/Pandas/141_150_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 141](#1)
* [Ćwiczenie 142](#2)
* [Ćwiczenie 143](#3)
* [Ćwiczenie 144](#4)
* [Ćwiczenie 145](#5)
* [Ćwiczenie 146](#6)
* [Ćwiczenie 147](#7)
* [Ćwiczenie 148](#8)
* [Ćwiczenie 149](#9)
* [Ćwiczenie 150](#10)

### <a name='0'></a>Import biblioteki

In [ ]:
import numpy as np
import pandas as pd

np.random.seed(42)
np.__version__

'1.25.2'

### <a name='1'></a> Ćwiczenie 141
Podane są dwa obiekty typu _Series_:

```
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
```
Połącz te dwa obiekty w jeden obiekt typu _DataFrame_ (jako dwie kolumny) i przypisz do zmiennej _df_. Nadaj nazwy kolumn odpowiednio _col1_ oraz _col2_.


In [ ]:
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))

df = pd.concat([s1, s2], axis=1)
df.columns = ['col1', 'col2']
df

col1      col2
0   0.892047 -0.846794
1   0.631139 -1.514847
2   0.794811 -0.446515
3   0.502637  0.856399
4   0.576904  0.214094
5   0.492518 -1.245739
6   0.195243  0.173181
7   0.722452  0.385317
8   0.280772 -0.883857
9   0.024316  0.153725
10  0.645472  0.058209
11  0.177111 -1.142970
12  0.940459  0.357787
13  0.953929  0.560785
14  0.914864  1.083051
15  0.370159  1.053802
16  0.015457 -1.377669
17  0.928319 -0.937825
18  0.428184  0.515035
19  0.966655  0.513786

### <a name='2'></a> Ćwiczenie 142
Wytnij z obiektu _df_ wiersze, dla których wartość w kolumnie _col2_ jest pomiędzy 0.0 a 1.0 (włącznie).

In [ ]:
df[(df['col2'] >= 0) & (df['col2'] <= 1)]

col1      col2
3   0.502637  0.856399
4   0.576904  0.214094
6   0.195243  0.173181
7   0.722452  0.385317
9   0.024316  0.153725
10  0.645472  0.058209
12  0.940459  0.357787
13  0.953929  0.560785
18  0.428184  0.515035
19  0.966655  0.513786

### <a name='3'></a> Ćwiczenie 143
Przypisz nową kolumnę o nazwie _col3_, która przyjmie wartość 1 w momencie gdy w kolumnie _col2_ jest wartość nieujemna i przeciwnie -1.

In [ ]:
df['col3'] = df['col2'].map(lambda x: 1 if x >= 0 else -1)
df

col1      col2  col3
0   0.892047 -0.846794    -1
1   0.631139 -1.514847    -1
2   0.794811 -0.446515    -1
3   0.502637  0.856399     1
4   0.576904  0.214094     1
5   0.492518 -1.245739    -1
6   0.195243  0.173181     1
7   0.722452  0.385317     1
8   0.280772 -0.883857    -1
9   0.024316  0.153725     1
10  0.645472  0.058209     1
11  0.177111 -1.142970    -1
12  0.940459  0.357787     1
13  0.953929  0.560785     1
14  0.914864  1.083051     1
15  0.370159  1.053802     1
16  0.015457 -1.377669    -1
17  0.928319 -0.937825    -1
18  0.428184  0.515035     1
19  0.966655  0.513786     1

### <a name='4'></a> Ćwiczenie 144
Przypisz nową kolumnę o nazwie _col4_, która przytnie wartości z kolumny _col2_ do przedziału $[-1.0, 1.0]$.
Innymi słowy, dla wartości poniżej -1.0 należy wstawić wartość -1.0, dla wartości powyżej 1.0 wartość 1.0.


In [ ]:
df['col4'] = df['col2'].clip(-1.0, 1.0)
df

col1      col2  col3      col4
0   0.892047 -0.846794    -1 -0.846794
1   0.631139 -1.514847    -1 -1.000000
2   0.794811 -0.446515    -1 -0.446515
3   0.502637  0.856399     1  0.856399
4   0.576904  0.214094     1  0.214094
5   0.492518 -1.245739    -1 -1.000000
6   0.195243  0.173181     1  0.173181
7   0.722452  0.385317     1  0.385317
8   0.280772 -0.883857    -1 -0.883857
9   0.024316  0.153725     1  0.153725
10  0.645472  0.058209     1  0.058209
11  0.177111 -1.142970    -1 -1.000000
12  0.940459  0.357787     1  0.357787
13  0.953929  0.560785     1  0.560785
14  0.914864  1.083051     1  1.000000
15  0.370159  1.053802     1  1.000000
16  0.015457 -1.377669    -1 -1.000000
17  0.928319 -0.937825    -1 -0.937825
18  0.428184  0.515035     1  0.515035
19  0.966655  0.513786     1  0.513786

### <a name='5'></a> Ćwiczenie 145
Znajdź 5 największych wartości dla kolumny _col2_.

In [ ]:
df['col2'].nlargest(5)

14    1.083051
15    1.053802
3     0.856399
13    0.560785
18    0.515035
Name: col2, dtype: float64

Znajdź 5 najmniejszych wartości dla kolumny _col2_.

In [ ]:
df['col2'].nsmallest(5)

1    -1.514847
16   -1.377669
5    -1.245739
11   -1.142970
17   -0.937825
Name: col2, dtype: float64

### <a name='6'></a> Ćwiczenie 146
Wyznacz skumulowaną sumę dla każdej kolumny.

__Wskazówka:__ Użyj metody _pd.DataFrame.cumsum()_.

In [ ]:
df.cumsum()

col1      col2  col3      col4
0    0.892047 -0.846794    -1 -0.846794
1    1.523185 -2.361641    -2 -1.846794
2    2.317996 -2.808156    -3 -2.293309
3    2.820634 -1.951757    -2 -1.436910
4    3.397537 -1.737663    -1 -1.222816
5    3.890055 -2.983402    -2 -2.222816
6    4.085298 -2.810221    -1 -2.049635
7    4.807750 -2.424904     0 -1.664318
8    5.088523 -3.308761    -1 -2.548175
9    5.112839 -3.155036     0 -2.394450
10   5.758311 -3.096827     1 -2.336241
11   5.935422 -4.239798     0 -3.336241
12   6.875880 -3.882010     1 -2.978454
13   7.829809 -3.321226     2 -2.417670
14   8.744673 -2.238175     3 -1.417670
15   9.114832 -1.184373     4 -0.417670
16   9.130288 -2.562042     3 -1.417670
17  10.058607 -3.499867     2 -2.355495
18  10.486791 -2.984832     3 -1.840459
19  11.453446 -2.471046     4 -1.326673

### <a name='7'></a> Ćwiczenie 147
Wyznacz medianę dla zmiennej _col2_ (inaczej kwantyl rzędu 0.5).

In [ ]:
df['col2'].median()

0.16345301589835495

### <a name='8'></a> Ćwiczenie 148
Wytnij wiersze, dla których zmienna _col2_ przyjmuje wartości większe od 0.0.

In [ ]:
df[df['col2'] > 0]

col1      col2  col3      col4
3   0.502637  0.856399     1  0.856399
4   0.576904  0.214094     1  0.214094
6   0.195243  0.173181     1  0.173181
7   0.722452  0.385317     1  0.385317
9   0.024316  0.153725     1  0.153725
10  0.645472  0.058209     1  0.058209
12  0.940459  0.357787     1  0.357787
13  0.953929  0.560785     1  0.560785
14  0.914864  1.083051     1  1.000000
15  0.370159  1.053802     1  1.000000
18  0.428184  0.515035     1  0.515035
19  0.966655  0.513786     1  0.513786

### <a name='9'></a> Ćwiczenie 149
Wytnij 5 pierwszych wierszy obiektu _df_ i przekształć je do słownika.

In [ ]:
df.head(5).to_dict()

{'col1': {0: 0.8920465551771133,
  1: 0.6311386259972629,
  2: 0.7948113035416484,
  3: 0.5026370931051921,
  4: 0.5769038846263591},
 'col2': {0: -0.846793718068405,
  1: -1.5148472246858646,
  2: -0.4465149520670211,
  3: 0.8563987943234723,
  4: 0.21409374413020396},
 'col3': {0: -1, 1: -1, 2: -1, 3: 1, 4: 1},
 'col4': {0: -0.846793718068405,
  1: -1.0,
  2: -0.4465149520670211,
  3: 0.8563987943234723,
  4: 0.21409374413020396}}

### <a name='10'></a> Ćwiczenie 150
Wytnij 5 pierwszych wierszy obiektu _df_, przekształć je do formatu Markdown i przypisz do zmiennej _df_markdown_.

In [ ]:
df_markdown = df.head().to_markdown()

Wydrukuj do konsoli zawartość zmiennej _df_markdown_.

In [ ]:
df_markdown

'|    |     col1 |      col2 |   col3 |      col4 |\n|---:|---------:|----------:|-------:|----------:|\n|  0 | 0.892047 | -0.846794 |     -1 | -0.846794 |\n|  1 | 0.631139 | -1.51485  |     -1 | -1        |\n|  2 | 0.794811 | -0.446515 |     -1 | -0.446515 |\n|  3 | 0.502637 |  0.856399 |      1 |  0.856399 |\n|  4 | 0.576904 |  0.214094 |      1 |  0.214094 |'

|    |     col1 |      col2 |   col3 |      col4 |
|---:|---------:|----------:|-------:|----------:|
|  0 | 0.892047 | -0.846794 |     -1 | -0.846794 |
|  1 | 0.631139 | -1.51485  |     -1 | -1        |
|  2 | 0.794811 | -0.446515 |     -1 | -0.446515 |
|  3 | 0.502637 |  0.856399 |      1 |  0.856399 |
|  4 | 0.576904 |  0.214094 |      1 |  0.214094 |

Skopiuj wynik uruchomienia powyższej komórki i wklej poniżej:

\# tutaj wpisz rozwiązanie
